src/utils/image-processor.js

```diff
@@ -91,90 +92,73 @@ const shrinkFile = function(file, settings) {
 }
 
 const shrinkImage = function(img, settings) {
-  var cx,
-    df,
-    h,
-    maxSize,
-    maxSquare,
-    originalW,
-    ratio,
-    run,
-    runNative,
-    sH,
-    sW,
-    step,
-    w
   // in -> image
   // out <- canvas
-  df = $.Deferred()
-  step = 0.71 // sohuld be > sqrt(0.5)
-  if (img.width * step * img.height * step < settings.size) {
-    return df.reject('not required')
-  }
-  sW = originalW = img.width
-  sH = img.height
-  ratio = sW / sH
-  w = Math.floor(Math.sqrt(settings.size * ratio))
-  h = Math.floor(settings.size / Math.sqrt(settings.size * ratio))
-  maxSquare = 5000000 // ios max canvas square
-  maxSize = 4096 // ie max canvas dimensions
-  run = function() {
-    if (sW <= w) {
-      df.resolve(img)
-      return
+  const df = $.Deferred()
+
+  const maxSizePromise = maxCanvasSize()
+  maxSizePromise.done(({ maxSize, maxSquare }) => {
```

Comment:
Since we have unlimited value in `settings.size`, we need to check it on every step. It's more convenient to make a wrapper around `maxCanvasSize` and get already limited values:

```js
limitedCanvasSize(w, h, size) {
  const df = $.Deferred()
  maxCanvasSize().done(({ maxSize, maxQuery }) => {
    ...
    df.resolve(w, h, size);
  })
}

limitedCanvasSize(w, h, settings.size).done(({ maxSize, settingsSize }) => {
...
})
```